### PR TITLE
LTR390 bugfix

### PR DESCRIPTION
**Pull request by amihajlovic**
# What does this implement/fix?

Implements combining data from three data registers as specified in the datasheet. 
https://optoelectronics.liteon.com/upload/download/DS86-2015-0004/LTR-390UV_Final_%20DS_V1%201.pdf
```
ALS_DATA Registers (Read Only)
//The registers 0x0D, 0x0E & 0x0F contain ALS data, up to 20bits.
Register_Addr = 0x0D // ALS_DATA_0 address
Data1 = ReadByte(Slave_Addr, Register_Addr)
Register_Addr = 0x0E // ALS_DATA_1 address
Data2 = ReadByte(Slave_Addr, Register_Addr)
Register_Addr = 0x0F // ALS_DATA_2 address
Data3 = ReadByte(Slave_Addr, Register_Addr) // ALS_DATA=Data3*65536+Data2*256+Data1.
UVS_DATA Registers (Read Only)
//The registers 0x10, 0x11 & 0x12 contain UVS data, up to 20bits.
Register_Addr = 0x10 // UVS_DATA_0 address
Data1 = ReadByte(Slave_Addr, Register_Addr)
Register_Addr = 0x11 // UVS_DATA_1 address
Data2 = ReadByte(Slave_Addr, Register_Addr)
Register_Addr = 0x12 // UVS_DATA_2 address
Data3 = ReadByte(Slave_Addr, Register_Addr) // UVS_DATA=Data3*65536+Data2*256+Data1.
``` 

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:
```yaml

i2c:
  frequency: 100000
   
logger:
  level: DEBUG

sensor:
  - platform: ltr390
    address: 0x53
    uv_index:
      id: ltr3901_uv_index
      name: "UV Index"
    
    light:
      id: ltr3901_lux
      name: "Light"     
      
    uv:
      id: ltr3901_uv_sensor_count
      name: "UV Sensor Counts"
      
    ambient_light:
      id: ltr3901_lux_sensor_count
      name: "Light Sensor Counts"
      
    gain: X18
    resolution: 20
    update_interval: 5s

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/ltr390/ltr390.cpp [hunk 1]
```diff
@@ -56,12 +47,20 @@ optional<uint32_t> LTR390Component::read_sensor_data_(LTR390MODE mode) {
...
+    msb_reg = LTR390_ALSDATA_MSB;
+    hsb_reg = LTR390_ALSDATA_HSB;
   }
-
-  return little_endian_bytes_to_int(buffer, num_bytes);
+  uint8_t lsb = this->reg(lsb_reg).get();
+  uint8_t msb = this->reg(msb_reg).get();
+  uint8_t hsb = this->reg(hsb_reg).get();
+  hsb &= 0x0F;
+  uint32_t out = ((uint32_t) hsb << 16) | ((uint16_t) msb << 8) | lsb;
```
Comment (jesserockz) [inline esphome/components/ltr390/ltr390.cpp:62]:
```suggestion
  uint32_t out = encode_uint24(hsb, msb, lsb);
```

Comment (jesserockz) [reply esphome/components/ltr390/ltr390.cpp:62]: This will need to also `#include "esphome/core/helpers.h";`

**Comment by amihajlovic** [comment]
I'm closing this for now. I've retested it during daytime, it's working well for UVI sensor, but ambient light is not working as expected now.

**Comment by sjtrny** [comment]
Maybe I'm missing something but what's the point of this? I just tested with

```
ESP_LOGD(TAG, "%u", lsb);
ESP_LOGD(TAG, "%u", msb);
ESP_LOGD(TAG, "%u", hsb);

ESP_LOGD(TAG, "%u", little_endian_bytes_to_int(buffer, num_bytes));
```

which gives me


```
[19:56:50][D][ltr390:083]: 143
[19:56:50][D][ltr390:084]: 1
[19:56:50][D][ltr390:085]: 0
[19:56:50][D][ltr390:087]: 399
```


which means the existing code and your new version give the same result since 143 + 256*1 = 399.

**Comment by amihajlovic** [comment]
> Maybe I'm missing something but what's the point of this? I just tested with
> 
> ```
> ESP_LOGD(TAG, "%u", lsb);
> ESP_LOGD(TAG, "%u", msb);
> ESP_LOGD(TAG, "%u", hsb);
> 
> ESP_LOGD(TAG, "%u", little_endian_bytes_to_int(buffer, num_bytes));
> ```
> 
> which gives me
> 
> ```
> [19:56:50][D][ltr390:083]: 143
> [19:56:50][D][ltr390:084]: 1
> [19:56:50][D][ltr390:085]: 0
> [19:56:50][D][ltr390:087]: 399
> ```
> 
> which means the existing code and your new version give the same result since 143 + 256*1 = 399.

You are correct. I've started investigating this because I'm getting some weird readouts while using this component, while I get "expected" results when using https://github.com/levkovigor/LTR390. Two things stood out immediately - I would get ~8700lux and UVI of <1. Local weather station reported UVI >4 at that time. 
At first it looked like this change solved the issue but while performing a more detailed test during daytime it would seem it hasn't (reason I've closed the PR).

Right now it looks like it's linked to using global gain and resolution settings instead of using different values for each sensor.  Specifically - when using the example code found in the levkovigor's repository I get readouts that are more in line with what I would expect - UVI >5, ~60000lux. And that code is using very different gain and resolution for UV and ALS readouts. 
I'll investigate this further and reopen the PR if there are some relevant changes to be made.

**Pull request closed without merging**